### ISSUE TITLE
Weighted Hash Embeddings

### ISSUE THREAD

**Issue by YannDubs** [explosion/thinc#107]
**TL;DR: Improving HashEmbed class by using a weighted second matrix.**

I was recently looking at the HashEmbed class (after seeing some old tweets from @honnibal ). I was wondering if there was any reason not to use a weighted second matrix as in [Hash Embeddings for Efficient Word Representations (NIPS 2017)](https://arxiv.org/abs/1709.03933)? 

The second matrix would be long but narrow (so not many parameters, usually same order of magnitude than first matric) and contains for each word the weights it should give for different hashes. E.g. "dog: [0.7, 0.2, 0.9]" such that word dog is not 0.7*hash_1(dog) + 0.2 * hash_2(dog)  + 0.9 * has_3(dog). I've implemented it for the NIPS implementation challenge in 2017 and it worked very well (https://github.com/YannDubs/Hash-Embeddings). Small drawing I made back then to explain the method:

<img width="613" alt="embeddings_explanation" src="https://user-images.githubusercontent.com/24327668/61459108-e75d0400-a963-11e9-93f8-20bdf006e20a.png">

I was thus wondering if there is any reason not to use that instead ? Have you tried it?

**Comment by honnibal** [comment]
Well, try it and see :). In my early experiments it didn't help, but maybe I implemented it wrongly or something else was different. I developed the hash embeddings separately from that paper --- I think a few different people had the same idea. If the weighting matrix helps I'd be very interested to have it included.

I don't actually see an experiment in their paper that evaluates the impact of the weighting dictionary, but maybe I haven't read closely enough and have missed it.

**Comment by YannDubs** [comment]
Thansk for the quick reply

No they don't compare unfortunately! a few papers and people came up with that idea in parallel (which makes sense as it is very natural). I was just wondering if you tried with the weights (as it is also natural), I might run a few experiments next month.

**Comment by honnibal** [comment]
I'll close this in the meantime :)
